File: built-in plugins dsp/mb-filter-lp.cpp
/**
 * MB Low-Pass Filter
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Resonant low-pass filter with multiple slope options
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FILTER_LP_H
#define MB_FILTER_LP_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFilterLp : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-filter-lp";
    static constexpr const char* PLUGIN_NAME    = "MB Low-Pass Filter";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float cutoff = 5000f;  // range [20, 20000]
    float resonance = 0.3f;  // range [0, 1]
    float slope = 12f;  // range [6, 48]
    float mix = 1f;  // range [0, 1]
    };

    MbFilterLp() = default;
    ~MbFilterLp() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.cutoff = std::clamp(params.cutoff, 20f, 20000f);
        params.resonance = std::clamp(params.resonance, 0f, 1f);
        params.slope = std::clamp(params.slope, 6f, 48f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Low-Pass Filter
        return input;
    }
};

#endif // MB_FILTER_LP_H

File: built-in plugins dsp/mb-filter-lp.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbFilterLpPlugin: PluginDefinition = { id: 'mb-filter-lp', slug: 'mb-filter-lp', name: 'MB Low-Pass Filter', category: 'effect', type: 'eq' as any, version: '1.0.0', description: 'Resonant low-pass filter with multiple slope options', author: 'Max Booster', parameters: [{ id: 'cutoff', name: 'Cutoff', type: 'float', defaultValue: 5000, minValue: 20, maxValue: 20000, automatable: true }, { id: 'resonance', name: 'Resonance', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }, { id: 'slope', name: 'Slope', type: 'float', defaultValue: 12, minValue: 6, maxValue: 48, automatable: false }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { cutoff: 5000, resonance: 0.3, slope: 12, mix: 1 } };

export default MbFilterLpPlugin;
